### DISCUSSION TITLE
[FEATURE REQUEST] Abitlity to make secured connections in pillar_ldap

### DISCUSSION THREAD

**Discussion by kevkha** [saltstack/salt#67515] [Feature Requests]
**Is your feature request related to a problem? Please describe.**
We need to get `ldaps` working but `salt.pillar.pillar_ldap` does not support `uri` from this [line of code](https://github.com/saltstack/salt/blob/master/salt/pillar/pillar_ldap.py#L274).

**Describe the solution you'd like**
if `uri` is defined in `pillar_ldap.yaml` it should override both `server` and `port`

```
schema = "ldaps" if tls else "ldap"
  if connargs["uri"] == ""
    connargs["uri"] = "{}://{}:{}".format(schema, connargs["server"], connargs["port"])
```
**Describe alternatives you've considered**
Alternately, change code to support `uri` only 

secured connections
```
acme-user:
  uri:       ldaps://auth.acme.com:636
  tls:       true
```
or unsecured connections
```
acme-user:
  uri:       ldap://auth.acme.com:389
  tls:       false
```
**Additional context**
Example of `pillar_ldap.yaml`
```
acme-user:
  uri:       ldaps://auth.acme.com:636
  server:    auth.acme.com
  port:      636
  tls:       true
```

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

**Comment by welcome** [comment]
Hi there! Welcome to the Salt Community! Thank you for making your first contribution. We have a lengthy process for issues and PRs. Someone from the Core Team will follow up as soon as possible. In the meantime, here’s some information that may help as you continue your Salt journey.
Please be sure to review our [Code of Conduct](https://github.com/saltstack/salt/blob/master/CODE_OF_CONDUCT.md). Also, check out some of our community resources including:

  - [Community Wiki](https://github.com/saltstack/community/wiki)
  - [Salt’s Contributor Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html)
  - [Join our Community Slack](https://join.slack.com/t/saltstackcommunity/shared_invite/zt-3av8jjyf-oBQ2M0vhXOhJpNpRkPWBvg)
  - [IRC on LiberaChat](https://web.libera.chat/#salt)
  - [Salt Project YouTube channel](https://www.youtube.com/channel/UCpveTIucFx9ljGelW63-BWg)
  - [Salt Project Twitch channel](https://www.twitch.tv/saltprojectoss)

There are lots of ways to get involved in our community. Every month, there are around a dozen opportunities to meet with other contributors and the Salt Core team and collaborate in real time. The best way to keep track is by subscribing to the Salt Community Events Calendar.
If you have additional questions, email us at saltproject@vmware.com. We’re glad you’ve joined our community and look forward to doing awesome things with you!

**Comment by Frizlab** [comment]
Fully agree, it’s very weird not to have `ldaps` support since a few years ago already, especially since LDAP with STARTTLS [should be considered a bad practice](https://unix.stackexchange.com/a/721979).

**Comment by Frizlab** [comment]
FWIW the patch could be as simple as
```patch
--- /usr/lib/python2.7/dist-packages/salt/pillar/pillar_ldap_old.py	2023-11-16 11:42:37.795149374 +0000
+++ /usr/lib/python2.7/dist-packages/salt/pillar/pillar_ldap.py	2023-11-16 11:43:24.423235480 +0000
@@ -274,7 +274,7 @@
     '''
     # Build LDAP connection args
     connargs = {}
-    for name in ['server', 'port', 'tls', 'binddn', 'bindpw', 'anonymous']:
+    for name in ['uri', 'server', 'port', 'tls', 'binddn', 'bindpw', 'anonymous']:
         connargs[name] = _config(name, conf)
     if connargs['binddn'] and connargs['bindpw']:
         connargs['anonymous'] = False
```
